pkg/proxy/endpoints.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+// EndpointsChangeMap maps a service to its endpointsChange.
+type EndpointsChangeMap struct {
+	// lock protects items.
+	lock sync.Mutex
+	// hostname is the host where kube-proxy is running.
+	hostname string
+	// items is a service to its endpointsChange map.
+	items map[types.NamespacedName]*endpointsChange
+}
+
+// endpointsChange contains all changes to endpoints that happened since proxy rules was synced.  For a single object,
+// changes are accumulated, i.e. previous is state from before all of them,
+// current is state after applying all of those.
+type endpointsChange struct {
+	previous EndpointsMap
+	current  EndpointsMap
+}
+
+// EndpointsMap maps a service to its endpoints.
+type EndpointsMap map[ServicePortName][]Endpoints
+
+// NewEndpointsChangeMap initialize an EndpointsChangeMap
+func NewEndpointsChangeMap(hostname string) EndpointsChangeMap {
+	return EndpointsChangeMap{
+		hostname: hostname,
+		items:    make(map[types.NamespacedName]*endpointsChange),
+	}
+}
+
+// UpdateEndpointMapResult is the updated results after applying endpoints changes.
+type UpdateEndpointMapResult struct {
+	// HCEndpointsLocalIPSize maps an endpoints name to the length of its local IPs.
+	HCEndpointsLocalIPSize map[types.NamespacedName]int
+	// StaleServiceNames identifies if an endpoints service pair is stale.
+	StaleEndpoints map[EndpointServicePair]bool
+	// StaleServiceNames identifies if a service is stale.
+	StaleServiceNames map[ServicePortName]bool
+}
+
+// UpdateEndpointsMap updates endpointsMap base on the given changes.
+// The changes map is cleared after applying them.
```

Comment:
In the comment you claim that the changes map is cleared in this function, but in reality the clearing happens in endpointsMap.apply. Can you move that portion of the comment to the apply function to make it more clear which piece is actually doing the clearing?

Reply:
Sounds fair, it's fixed now.